Specification for standard lightwallet client/server API

ndorf opened the pull request:
As implemented by OpenMonero, MyMonero and [the Monero Project](https://github.com/monero-project/monero/pull/4139).

moneroexamples commented:
Are there updates regarding ringct conbase txs in `rct` field and making all xmr amounts string?

ndorf commented:
Sure thing. I also renamed the per_kb_field as discussed earlier, fixed the field descriptions as [suggested by stoffu](https://github.com/monero-project/meta/pull/272#discussion_r225030974), and added a new field `fee_mask` to pass the fee quantization mask to the client.

The changes are smaller [with whitespace ignored](https://github.com/monero-project/meta/pull/287/commits/e08a08695834af007afbbd0f99053ed1c5e7d7e3?utf8=%E2%9C%93&diff=unified&w=1). You can also see them as individual commits on [this otherwise-identical branch](https://github.com/ndorf/meta/commits/lightwallet_api_1).

ndorf commented:
> mymonero/mymonero-app-js#240

Yes, the `is_rct` flag is meant to fix that issue, as [suggested by luigi therein](https://github.com/mymonero/mymonero-app-js/issues/240#issuecomment-436111160). When `is_rct == true && rct.length == 0`, the output is from a ringct coinbase tx and the unencrypted identity mask must be used. That should be all that's needed, unless I'm misunderstanding something?

> Also what is fee_mask?

I asked about that [a few days ago](https://github.com/monero-project/meta/pull/272#discussion_r231698178). Basically, the final fee should be rounded to be a multiple of this value when constructing the tx. Historically, it has been hardcoded and has never changed; however, with [mooo's v8 changes](https://github.com/monero-project/monero/commit/5ffb2ff9b7c301eda5811a939c705f26627c4735), the mainline wallet has been updated to request it from the daemon. Since that means that, at least theoretically, it can now change at any time, it seems to make sense to support that for the lightwallet client as well. The value should be retrieved from `monerod` using the `get_fee_quantization_mask` RPC call. What do you think?

>  I think we already moved to gamma distribution 

Good catch, I'll change this. Thanks.

ndorf commented:
Oops, I seem to have missed the other half of that discussion. Ignore the `is_rct` portion for a moment, please

ndorf commented:
I've now replaced the `is_rct` field with a note stating stating that "Coinbase ringct transactions shall have the identity mask and a zero amount in [the `rct` field]," in commit https://github.com/ndorf/meta/commit/16aa96d3f449c1fc5f620f894cf741cf32643e15 and squashed into this PR.

I don't think anything else is required to correctly handle these txes, is it? Does the usage need to be clarified further?

ndorf commented:
Actually, it looks like the `rct` field description wasn't correct; it stated that it contained "decrypted" mask and amount values, but in reality both openmonero and mymonero appear to send the raw, encrypted values (except for coinbase txes).

So I replaced it with a new one (in commit https://github.com/monero-project/meta/commit/37689cec3e591ab2779b26ed4e202bffda02bbe9, and [squashed](https://github.com/monero-project/meta/pull/287/commits/00e59760ec05d1afec43186d54bb5859d617c10a?utf8=%E2%9C%93&diff=unified&w=1)) that reads as follows:

> `rct` is, for ringct outputs, a 96-byte blob containing the concatenation of
> the public commitment, then the ringct mask value, and finally the ringct
> amount value. For coinbase transactions, the mask is always the identity mask
> and the amount is zero; for other transactions, the mask and amount are the
> respective raw encrypted values, which must be decrypted by the client using
> the view secret key. For non-ringct outputs, this field is nil.

moneroexamples commented:
Thanks for explanations and changes.

`For coinbase transactions` should be `For ringct coinbase transactions` . Non-ringct coinbase txs would be treated normally as other non-ringct txs.

Thx for  info about and links and quantization_mask. I see it is also in the official monero wallet. Missed that before. 

For now I don't have any other comments. I'm sure something new will pop up as time progresses:-). Probably the only other comment I would have would be to have some actual examples of all these api queries, rct fields, quantization masks, error responses, etc. But this can be done later, or viewed through openmonero  implementation.

Anyway, maybe @vtnerd  or @paulshapiro want to comment more, especially about the `rct` field.

paulshapiro commented:
Yep, that looks good to me.

moneroexamples commented:
@ndorf 

Sounds good to me.

vtnerd commented on api/lightwallet_rest.md:
```diff
@@ -309,11 +318,12 @@ was actually spent.
+|    Field     |        Type       |                Description              |
+|--------------|-------------------|-----------------------------------------|
+| per_byte_fee | uint64-string     | Estimated network fee                   |
+| fee_mask     | uint64-string     | Fee quantization mask                   |
```
This can _probably_ be a `uint64` field, because it is unlikely that this mask will exceed 2^53. This would be masking 1000 XMR or so.

vtnerd commented on api/lightwallet_rest.md:
```diff
@@ -309,11 +318,12 @@ was actually spent.
-| outputs    | array of output's | Outputs possibly available for spending |
+|    Field     |        Type       |                Description              |
+|--------------|-------------------|-----------------------------------------|
+| per_byte_fee | uint64-string     | Estimated network fee                   |
```
This field is currently a `uint64` for the same reason, this was not a typo.

ndorf replied:
Yeah, @moneroexamples requested all XMR amount fields to be `uint64-string` for consistency. It seemed reasonable enough, assuming the costs of parsing the string are negligible. Is that not the case?

vtnerd commented on api/lightwallet_rest.md:
```diff
@@ -76,9 +82,12 @@ Information needed to spend an output.
-> is returned as a single 96-byte binary blob.
+> `rct` is, for ringct outputs, a 96-byte blob containing the concatenation of
+> the public commitment, then the ringct mask value, and finally the ringct
+> amount value. For ringct coinbase outputs, the mask is always the identity
```
This comment got lost I think ...

There are no "other ringct outputs", this object is strictly for spending an output. The `amount` field would have to be updated otherwise; ringct would always be zero for "other ringct outputs".

Also, the ringct values are always encrypted.

ndorf replied:
"Other ringct outputs" here means other than the "ringct coinbase outputs" mentioned earlier, i.e. non-coinbase ringct outputs. I guess I should just write exactly that, for clarity.

For the coinbase txes, the server is now to construct a special rct object consisting of the 
_unencrypted_ values `rct::identity()` for the mask and `0` for the amount. Previously, the client was [unable to spend ringct coinbase outputs](https://github.com/mymonero/mymonero-app-js/issues/240) at all. This was the solution chosen [therein](https://github.com/mymonero/mymonero-app-js/issues/240#issuecomment-436352783), not least because the OpenMonero server already does it this way. 

Does this make sense now, or do you think something (other than `s/other/non-coinbase/`) still needs to be changed?

vtnerd commented on api/lightwallet_rest.md:
```diff
@@ -0,0 +1,401 @@
+> If clients are creating multiple rings with the same amount, they must set
+> `count` to the mixin level and add the value to `amounts` multiple times.
+> Server must respond to each value in `amounts`, even if the value appears
+> multiple times.gi
```
Some weird typo I left (`gi`).

ndorf replied:
Found the vi user :)

ndorf commented:
Updated with typo fix (1018025a44ca5622910b7f9ba019df2565aa46ec) and "non-coinbase" clarification (4a7ef32a52cf206da027125341b490e73bad79b1).

serhack commented on api/lightwallet_rest.md:
```diff
@@ -0,0 +1,402 @@
+| RUB * | float32 | RUB/XMR exchange rate |
+| ZAR * | float32 | ZAR/XMR exchange rate |
+
+> If an exchange rate is unavailable, the server field shall omit the field
```
I think it could be useful and helpful to mention from where Monero software will take the exchange rate.

vtnerd replied:
Does it need to be from a consistent place for all implementations? This seemed restrictive.

serhack replied:
I was trying to ask: is there any central source for retrieving prices? I saw you put cryptocompare API in your pull request on monero-project/monero.

ndorf replied:
I think the source of exchange rates has to remain up to the implementation. There doesn't exist a universally canonical source, and I don't see any reason to force implementations to use the same ones (or even provide them at all -- note, these fields are marked as optional.)

serhack replied:
I agree with you.

serhack commented on api/lightwallet_rest.md:
```diff
@@ -0,0 +1,402 @@
+| address   | base58-address | Address to retrieve                   |
+| view_key  | binary         | View key bytes for authorization      |
+
+> If `address` is not authorized, the server must return a HTTP 405
```
The HTTP code for Forbidden error is 403. (405 - Method not allowed). Should we keep 405 or 403?

vtnerd replied:
This was supposed to be 403, but I put 405 got in my head for some reason. The open source PR is using 403 currently.

serhack replied:
Thanks, I did not understand what PR I should review :-)

serhack commented on api/lightwallet_rest.md:
```diff
@@ -0,0 +1,402 @@
+| address  | base58-address | Address to retrieve                   |
+| view_key | binary         | View key bytes for authorization      |
+
+> If `address` is not authorized, the server must return a HTTP 405
```
I think it could be more useful HTTP 403.

Details about HTTP 403 RFC 2616 :

> The server understood the request, but is refusing to fulfill it. Authorization will not help and the request SHOULD NOT be repeated. If the request method was not HEAD and the server wishes to make public why the request has not been fulfilled, it SHOULD describe the reason for the refusal in the entity. If the server does not wish to make this information available to the client, the status code 404 (Not Found) can be used instead.

serhack commented on api/lightwallet_rest.md:
```diff
@@ -0,0 +1,402 @@
+> error must be returned.
+
+> If approval process is manual, a successful HTTP 200 OK and response object
+> must be returned. Subsequent requests shall be HTTP 405 "Forbidden" until
```
Should we keep 405 or 403?

vtnerd replied:
403 again.

serhack commented on api/lightwallet_rest.md:
```diff
@@ -0,0 +1,402 @@
+|        Field        |  Type   |            Description             |
+|---------------------|---------|------------------------------------|
+| new_address         | boolean | Whether account was just created   |
+| generated_locally * | boolean | Flag from initial account creation |
```
If I understand well, "*" means that could be null. An example: if the "generated_locally" is null, will the JSON response contain the value `"generated_locally"` as `""` ?

serhack commented on api/lightwallet_rest.md:
```diff
@@ -0,0 +1,402 @@
+|        Field        |  Type   |            Description             |
+|---------------------|---------|------------------------------------|
+| new_address         | boolean | Whether account was just created   |
+| generated_locally * | boolean | Flag from initial account creation |
```
What does `*` mean in this context?

serhack replied:
What is the purpose of having a response field as optional?

vtnerd replied:
It is not required for the server to implement the feature. Not sure if this is the best approach, but that was the idea (this was added very recently to mymonero and openmonero has (had?) no such concept).

ndorf commented:
Replaced "Forbidden" code with the correct 403.

moneroexamples commented:
Regarding `submit_raw_tx`

Its unclear what happens when submitting the raw tx data to deamon fails? What html error code should be returned, and how to provide info on why it failed. Current documentation does not provide any info about this.

While working on openmonero, I found that mostly these are the reason why commit fails: https://github.com/monero-project/monero/blob/0fccc78e3966477309cb4e9c037c0a9f004ea99e/src/rpc/core_rpc_server.cpp#L706-L722

moneroexamples commented:
@ndorf 

Thanks. 422 seems good. So probably this should be added to api docs.

ndorf commented:
@moneroexamples any opinion on the content of the error response? Is `{"error":"string"}` good enough?

moneroexamples commented:
MyMonero is using `Error` (capital letter), view-source:https://mymonero.com/js/controllers/send_coins.js?9 : 

```JavaScript
SubmitRawTx(parameters, fn)
    {
        const self = this
        const endpointPath = 'submit_raw_tx'
        self.$http.post(config.apiUrl + endpointPath, parameters).success(
            function(data)
            {
                fn(null, data)
            }
        ).error(
            function(data)
            {
                fn(data && data.Error ? data.Error : "Something went wrong while submitting your transaction");
            }
        );
        const requestHandle = 
        {
            abort: function()
            {
                console.warn("TODO: abort!")
            }
        }
        return requestHandle
    }
}
```
They use `Error` for other queries as well, e.g., https://mymonero.com/js/controllers/import_wallet.js?2

So I did same in OpenMonero: https://github.com/moneroexamples/openmonero/blob/devel/src/YourMoneroRequests.cpp#L964

moneroexamples commented:
Which API documentation and discussion is the current, most up to date? At the moment there are three resources about this:

1. on mymonero repo: https://github.com/mymonero/hosted-monero-api-spec
2. Previous PR to meta repo: https://github.com/monero-project/meta/pull/272
3. This PR to meto repo I use now: https://github.com/monero-project/meta/pull/287

My Guess is that no. 3 is the most resent one, thus maybe it would be better to close at lease no. 2? And I also understand that no. 1 is deprecated?

paulshapiro commented:
Correct on all counts. Let's get this PR merged as well (pending any final modifications) so that we can start using it as it was designed by making PRs to it.

moneroexamples commented:
@paulshapiro 

Thanks for quick answer. I have related question, and was not sure where I should post it. Thus my confusion and previous post. So here it is:

MyMonero is using `import_wallet_request` endpoint for importing accounts: https://github.com/mymonero/mymonero-app-js/blob/fe8028556d4582d1e747e52871e068fd19b886ee/local_modules/HostedMoneroAPIClient/HostedMoneroAPIClient_Base.js#L268. 

However, the API docs have `import_request` instead (https://github.com/monero-project/meta/blob/c139cb39ea56441fe7cb7ff13b74fe34b394f2f5/api/lightwallet_rest.md#import_request) assuming this is the current one?  So can I ask why different naming (a mistake maybe?) and also are there any other differences between api and the mymonero-app-js uses?

moneroexamples commented:
Also regarding `import_request` or just general importing accounts.

https://github.com/mymonero/mymonero-app-js/blob/fe8028556d4582d1e747e52871e068fd19b886ee/local_modules/HostedMoneroAPIClient/HostedMoneroAPIClient_Base.js#L268

The API suggest using legacy, plain text payments IDs. Such payments IDs should be deprecated and replaced with integrated address (what openmonero is already using) or just moved to sub-addresses. 

There is already a list made of services stuck with legacy payments ids (https://old.reddit.com/r/Monero/comments/abyivp/help_wanted_list_of_services_that_use_payment_ids/), and I don't think the API should encourge the use legacy payments ids. 

Also having subaddress for import_request would also simply the api.

paulshapiro commented:
@moneroexamples we're planning to move to a subaddress for the import endpoint, but I can't give a concrete timeline on which it'll happen. Hopefully within the next couple months.

paulshapiro commented:
@luigi1111 can we have this PR merged so we can begin using the document as it's intended, making PRs to it?

moneroexamples commented:
@paulshapiro @luigi1111 

I think API description in its current form is not complete. 

I think it would be better to merge it once its fully complete, and then future modification to it be about new features only, such as adding sub-addresses.

luigi1111 commented:
I'm fine with either.

paulshapiro commented:
@moneroexamples what's missing?

moneroexamples commented:
@paulshapiro 

For example, My questions regarding discrepancies between the API and MyMonero's handling of `import_request` haven't been addressed: https://github.com/monero-project/meta/pull/287#issuecomment-454202994

paulshapiro commented:
@moneroexamples ok, we'll get to answering that. 

As you say "for example", can you provide a complete list of what you see as the remaining items?

moneroexamples commented:
@paulshapiro 
I don't have complete list, as I find the issues as I go. So far I found the issue with `import_request` and earlier with error handling https://github.com/monero-project/meta/pull/287#issuecomment-451329581 and  https://github.com/monero-project/meta/pull/287#issuecomment-452143279 . With the error, was that mymonero is (or was at that time) using `Error`, instead or `error` as api states, and the error codes were not clear.

The time when I reported the issue with `import_request` was basically last time I looked into it. The `Error` and `import_request` questions haven't been answered. So I moved on to other non-API and mymonero related things. So I don't know if something has changed in mymonero regarding the issues?

paulshapiro commented:
Hi @moneroexamples,

let's remove the import request endpoint from this shared API spec. The OSS lightwallet server PR'd to Monero core doesn't use it either - it's done manually - and so it's a MyMonero specific thing. Our mistake. 

yeah I want to add a developer mode toggle to the MyMonero app to expose such features for developers but we're very resource constrained right now and I haven't had anyone pop up to talk with me about adding it. The app is for ordinary people rather than devs so it has to be designed properly to avoid confusion. For now the flag can be switched in one location at least.

paulshapiro commented:
I'll come back to you about the Error/error issue soon.

moneroexamples commented:
@paulshapiro 

Thanks for clarification. Sure, the `import_request` can be removed from the api. Alternatively, entire import_request api endpoint could be made optional.

Also thanks for looking into Error/error. 

Will try to check mymonero-app-js soon and see how the new flag for stagenet and testnet works.

p.s. Fully understand that developing mymonero is time consuming. I also lag behind with some issues in openmonero, onion explorer, and other monero-related projects.

ndorf commented:
Heads up for @vtnerd, we want to use a new format for the `output.rct` hex blob field (in the `get_unspent_outs`) response. [Details](https://github.com/mymonero/mymonero-app-js/issues/277#issuecomment-469395825):

>     1. null/undefined/empty: non-RingCT output (i.e, from version 1 tx)
> 
>     2. string "coinbase" (length 8): RingCT coinbase output
> 
>     3. string length 192: non-coinbase RingCT version 1 output with 256-bit amount and mask
> 
>     4. string length 80: non-coinbase RingCT version 2 output with 64-bit amount and no mask

ndorf commented:
Also, I'd like to propose three new fields in the `get_unspent_outs` response: `block_height`, `next_fork_block_height` and `next_fork_version`.

These would be for use by the client to determine whether or not to use, for example, RingCT type 4 when constructing new transactions.

ndorf commented:
> Also, I'd like to propose three new fields in the `get_unspent_outs` response: `block_height`, `next_fork_block_height` and `next_fork_version`.

After some coffee, I realized that only one field is needed, `fork_version` (the current one).

moneroexamples commented:
Wonder what is the status of subaddresess? 

Also has there aby any attempt at updateing the api documentation to reflect changes regarding BP2, mostly discussed here: https://github.com/mymonero/mymonero-app-js/issues/277

paulshapiro commented:
Hey @moneroexamples, 

1. Re subaddresses yes we'll support this but in my opinion, we cannot put the burden of all possible future features on this PR … this needs to be a living document to which we make PRs, in order to plan the future of the API. Each matter can then have, generally, its own contained thread. 

2. Yes, but feel free to do so as well. We'll suggest some improvements to the API to reflect the rct discrepancy across implementations but as of now, the client is hardened to it as far as I know, and so I think again, this is a future API improvement that we can justify leaving until after the PR is merged. As a matter of fact, it'll probably be easier to motivate to fix it afterwards since we'll finally have a reference spec for the API and people will be able to start using it, incentivizing them to adopt vtnerd's code as well. For now, this is not an issue, in my opinion.

We need to merge this PR imo. @vtnerd or @ndorf or @luigi1111 … Is there anything left outstanding from your povs?

moneroexamples commented:
@paulshapiro 

Thanks for clarification. I don't have further questions for now.
